r1cs/gadgets/crypto/src/prf/ripemd160.rs

```diff
@@ -0,0 +1,458 @@
+        },
+        // f(j, a, b, c) = (a AND b) OR (NOT(a) AND c)  (16 <= j <= 31)
+        // Note: It's the same as the sha256_ch function e.g. (a AND b) XOR (NOT(a) AND c)
+        // since the two logical expressions have the same truth table
+        round_idx if round_idx >= 16 && round_idx <= 31 => {
+            (
+                |a, b, c| (a & b) | ((!a) & c),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("ch {}", i)), a, b, c)
+            )
+        },
+        // f(j, a, b, c) = (a OR NOT(b)) XOR c (32 <= j <= 47)
+        round_idx if round_idx >= 32 && round_idx <= 47 => {
+            (
+                |a, b, c| (a | (!b)) ^ c,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("A OR NOT B {}", i)), &a, &b.not())?;
+                    Boolean::xor(cs.ns(|| format!("(A OR NOT B) XOR C {}", i)), &t, &c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (c AND a) OR (NOT(c) AND b)  (48 <= j <= 63)
+        // Note: It's the same as the second round function, but with permuted variables
+        round_idx if round_idx >= 48 && round_idx <= 63 => {
+            (
+                |a, b, c| (c & a) | ((!c) & b),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("permuted ch {}", i)), c, a, b)
+            )
+        },
+        // f(j, a, b, c) = (b OR NOT(c)) XOR a (64 <= j <= 79)
+        // Note: It's the same as the third round function but with permuted variables
+        round_idx if round_idx >= 64 && round_idx <= 79 => {
+            (
+                |a, b, c| (b | (!c)) ^ a,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("B OR NOT C {}", i)), &b, &c.not())?;
+                    Boolean::xor(cs.ns(|| format!("(B OR NOT C) XOR a {}", i)), &t, &a)
+                }
+            )
+        },
+        _ => unreachable!()
+    };
+
+    triop(cs, a, b, c, tri_fn, circ_fn)
+}
+
+pub fn ripemd160_block_no_padding<ConstraintF, CS>(
+    mut cs: CS,
+    input: &[Boolean],
+) -> Result<Vec<Boolean>, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>,
+{
+    assert_eq!(input.len(), 512);
+
+    Ok(
+        ripemd160_compression_function(&mut cs, &input, &get_ripemd160_iv())?
+            .into_iter()
+            .flat_map(|e| e.to_bits_le())
+            .collect(),
+    )
+}
+
+pub fn ripemd160<ConstraintF, CS>(mut cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    // append 0x80
+    padded.append(&mut UInt8::constant(1).into_bits_be());
+
+    // append K '0' bits, where K is the minimum number >= 0 such that L + 1 + K + 64 is a multiple of 512
+    while (padded.len() + 64) % 512 != 0 {
```

Comment:
Here the explanation is not precise. What is L? I guess that it is the length of the original message.
Apart from this, to me it looks that you are doing like this:
1)You first append the sequence 10 000 000, representing 0x80;
2) You add 0 until the new length is ≡ -64 mod 512;
3) You add the bits of L in little endian.

Reply:
Ulrich's proposal for the comments:
// append 0x80 = 1000 0000_2
    padded.append(&mut UInt8::constant(1).into_bits_be()); // 

    // append the remaining '0' bits up to the next multiple of 512, minus 64.
    while (padded.len() + 64) % 512 != 0 {
        padded.push(Boolean::constant(false));
    }
    // fill up to the multiple of 512 bits by appending the length of the unpadded 
    // message as a 64-bit little-endian integer.

Reply:
RIPEMD160 uses the same padding as SHA256 -> https://datatracker.ietf.org/doc/html/rfc6234#section-4.1